File: Schedule.cs
// Schedule.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using System.IO;
using System.Linq;

namespace TrainSchedule
{
    class Connection
    {
        public FromTo from { get; set; }
        public FromTo to { get; set; }
        public string duration { get; set; }
        public int transfers { get; set; }
    }

    class FromTo
    {
        public string departure { get; set; }
        public string arrival { get; set; }
        public Station station { get; set; }
    }

    class Station
    {
        public string name { get; set; }
    }

    class Response
    {
        public List<Connection> connections { get; set; }
    }

    class Program
    {
        private static readonly HttpClient client = new HttpClient();
        private static readonly Dictionary<string, (List<Connection> data, DateTime timestamp)> cache = new Dictionary<string, (List<Connection>, DateTime)>();
        private const int CacheTTL = 60;

        static async Task<List<Connection>> FetchConnections(string from, string to, string date, string time)
        {
            string key = $"{from}:{to}:{date}:{time}";
            if (cache.TryGetValue(key, out var entry))
            {
                if ((DateTime.Now - entry.timestamp).TotalSeconds < CacheTTL)
                    return entry.data;
            }
            string url = "https://transport.opendata.ch/v1/connections";
            var query = System.Web.HttpUtility.ParseQueryString(string.Empty);
            query["from"] = from;
            query["to"] = to;
            if (!string.IsNullOrEmpty(date)) query["date"] = date;
            if (!string.IsNullOrEmpty(time)) query["time"] = time;
            string fullUrl = url + "?" + query.ToString();
            try
            {
                var response = await client.GetStringAsync(fullUrl);
                var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
                var data = JsonSerializer.Deserialize<Response>(response, options);
                var connections = data?.connections ?? new List<Connection>();
                cache[key] = (connections, DateTime.Now);
                return connections;
            }
            catch (Exception e)
            {
                Console.WriteLine($"Error: {e.Message}");
                return new List<Connection>();
            }
        }

        static void Display(List<Connection> connections, int limit = 10, string sortBy = "departure")
        {
            if (connections == null || connections.Count == 0)
            {
                Console.WriteLine("No connections found.");
                return;
            }
            IEnumerable<Connection> sorted = connections;
            if (sortBy == "duration")
                sorted = sorted.OrderBy(c => c.duration);
            else if (sortBy == "changes")
                sorted = sorted.OrderBy(c => c.transfers);
            else
                sorted = sorted.OrderBy(c => c.from.departure);
            var list = sorted.ToList();
            Console.WriteLine($"\nFound {list.Count} connections.");
            Console.WriteLine(new string('-', 80));
            Console.WriteLine($"{"#",-3} {"Departure",-20} {"Arrival",-20} {"Duration",-10} {"Changes",-8}");
            for (int i = 0; i < Math.Min(list.Count, limit); i++)
            {
                var c = list[i];
                string depTime = c.from.departure?.Length > 5 ? c.from.departure.Substring(11, 5) : "?";
                string arrTime = c.to.arrival?.Length > 5 ? c.to.arrival.Substring(11, 5) : "?";
                string depSt = c.from.station?.name ?? "";
                string arrSt = c.to.station?.name ?? "";
                Console.WriteLine($"{i+1,-3} {depTime} {depSt,10} {arrTime} {arrSt,10} {c.duration,10} {c.transfers,8}");
            }
            if (list.Count > limit)
                Console.WriteLine($"... and {list.Count - limit} more");
        }

        static void ExportJSON(List<Connection> connections, string filename)
        {
            var options = new JsonSerializerOptions { WriteIndented = true };
            string json = JsonSerializer.Serialize(connections, options);
            File.WriteAllText(filename, json);
            Console.WriteLine($"Exported to {filename}");
        }

        static void ExportCSV(List<Connection> connections, string filename)
        {
            using var writer = new StreamWriter(filename);
            writer.WriteLine("Departure,Departure Station,Arrival,Arrival Station,Duration,Changes");
            foreach (var c in connections)
            {
                writer.WriteLine($"{c.from.departure},\"{c.from.station?.name}\",{c.to.arrival},\"{c.to.station?.name}\",{c.duration},{c.transfers}");
            }
            Console.WriteLine($"Exported to {filename}");
        }

        static async Task Interactive()
        {
            Console.WriteLine("=== Train Schedule ===");
            Console.Write("From station: ");
            string from = Console.ReadLine()?.Trim();
            if (string.IsNullOrEmpty(from)) { Console.WriteLine("Station required."); return; }
            Console.Write("To station: ");
            string to = Console.ReadLine()?.Trim();
            if (string.IsNullOrEmpty(to)) { Console.WriteLine("Station required."); return; }
            Console.Write("Date (YYYY-MM-DD, leave blank for today): ");
            string date = Console.ReadLine()?.Trim();
            if (string.IsNullOrEmpty(date)) date = DateTime.Now.ToString("yyyy-MM-dd");
            Console.Write("Time (HH:MM, leave blank for now): ");
            string time = Console.ReadLine()?.Trim();
            if (string.IsNullOrEmpty(time)) time = DateTime.Now.ToString("HH:mm");
            var conns = await FetchConnections(from, to, date, time);
            Display(conns);
            if (conns.Count > 0)
            {
                Console.Write("\nExport results? (y/n): ");
                string export = Console.ReadLine()?.Trim().ToLower();
                if (export == "y")
                {
                    Console.Write("Format (json/csv): ");
                    string fmt = Console.ReadLine()?.Trim().ToLower();
                    Console.Write("Filename (default: schedule." + fmt + "): ");
                    string fname = Console.ReadLine()?.Trim();
                    if (string.IsNullOrEmpty(fname)) fname = "schedule." + fmt;
                    if (fmt == "json") ExportJSON(conns, fname);
                    else if (fmt == "csv") ExportCSV(conns, fname);
                    else Console.WriteLine("Unknown format.");
                }
            }
        }

        static async Task Main(string[] args)
        {
            if (args.Length > 0)
            {
                // CLI mode - parse arguments
                var dict = new Dictionary<string, string>();
                for (int i = 0; i < args.Length; i++)
                {
                    if (args[i].StartsWith("--"))
                    {
                        string key = args[i].Substring(2);
                        if (i + 1 < args.Length && !args[i + 1].StartsWith("--"))
                            dict[key] = args[++i];
                        else
                            dict[key] = "true";
                    }
                }
                if (!dict.ContainsKey("from") || !dict.ContainsKey("to"))
                {
                    Console.WriteLine("--from and --to are required.");
                    return;
                }
                string from = dict["from"];
                string to = dict["to"];
                string date = dict.ContainsKey("date") ? dict["date"] : null;
                string time = dict.ContainsKey("time") ? dict["time"] : null;
                int limit = dict.ContainsKey("limit") ? int.Parse(dict["limit"]) : 10;
                string sortBy = dict.ContainsKey("sort") ? dict["sort"] : "departure";
                string exportFmt = dict.ContainsKey("export") ? dict["export"] : null;
                string output = dict.ContainsKey("output") ? dict["output"] : null;
                var conns = await FetchConnections(from, to, date, time);
                Display(conns, limit, sortBy);
                if (!string.IsNullOrEmpty(exportFmt) && conns.Count > 0)
                {
                    string fname = output ?? $"schedule.{exportFmt}";
                    if (exportFmt == "json") ExportJSON(conns, fname);
                    else if (exportFmt == "csv") ExportCSV(conns, fname);
                }
            }
            else
            {
                await Interactive();
            }
        }
    }
}

File: Schedule.java
// Schedule.java
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;
import com.google.gson.*;

public class Schedule {
    private static final String API_URL = "https://transport.opendata.ch/v1/connections";
    private static final int CACHE_TTL = 60;
    private static final Map<String, CachedData> cache = new HashMap<>();

    static class CachedData {
        List<Connection> data;
        long timestamp;
        CachedData(List<Connection> d, long ts) { data = d; timestamp = ts; }
    }

    static class Connection {
        FromTo from;
        FromTo to;
        String duration;
        int transfers;
    }

    static class FromTo {
        String departure;
        String arrival;
        Station station;
    }

    static class Station {
        String name;
    }

    static class Response {
        List<Connection> connections;
    }

    private static List<Connection> fetchConnections(String from, String to, String date, String time) throws Exception {
        String key = String.format("%s:%s:%s:%s", from, to, date, time);
        CachedData cached = cache.get(key);
        if (cached != null) {
            if ((System.currentTimeMillis() - cached.timestamp) / 1000 < CACHE_TTL) {
                return cached.data;
            }
        }
        StringBuilder urlBuilder = new StringBuilder(API_URL);
        urlBuilder.append("?from=").append(URLEncoder.encode(from, "UTF-8"));
        urlBuilder.append("&to=").append(URLEncoder.encode(to, "UTF-8"));
        if (date != null && !date.isEmpty()) urlBuilder.append("&date=").append(URLEncoder.encode(date, "UTF-8"));
        if (time != null && !time.isEmpty()) urlBuilder.append("&time=").append(URLEncoder.encode(time, "UTF-8"));
        URL url = new URL(urlBuilder.toString());
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setConnectTimeout(5000);
        conn.setReadTimeout(5000);
        if (conn.getResponseCode() != 200) {
            throw new RuntimeException("HTTP error: " + conn.getResponseCode());
        }
        try (BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream()))) {
            String json = br.lines().collect(Collectors.joining());
            Gson gson = new Gson();
            Response resp = gson.fromJson(json, Response.class);
            List<Connection> connections = resp.connections != null ? resp.connections : new ArrayList<>();
            cache.put(key, new CachedData(connections, System.currentTimeMillis()));
            return connections;
        }
    }

    private static void display(List<Connection> connections, int limit, String sortBy) {
        if (connections == null || connections.isEmpty()) {
            System.out.println("No connections found.");
            return;
        }
        List<Connection> sorted = new ArrayList<>(connections);
        if ("duration".equals(sortBy)) {
            sorted.sort(Comparator.comparing(c -> c.duration));
        } else if ("changes".equals(sortBy)) {
            sorted.sort(Comparator.comparingInt(c -> c.transfers));
        } else {
            sorted.sort(Comparator.comparing(c -> c.from.departure));
        }
        System.out.printf("%nFound %d connections.%n", sorted.size());
        System.out.println(new String(new char[80]).replace('\0', '-'));
        System.out.printf("%-3s %-20s %-20s %-10s %-8s%n", "#", "Departure", "Arrival", "Duration", "Changes");
        for (int i = 0; i < Math.min(sorted.size(), limit); i++) {
            Connection c = sorted.get(i);
            String depTime = c.from.departure != null && c.from.departure.length() > 5 ? c.from.departure.substring(11, 16) : "?";
            String arrTime = c.to.arrival != null && c.to.arrival.length() > 5 ? c.to.arrival.substring(11, 16) : "?";
            String depSt = c.from.station != null ? c.from.station.name : "";
            String arrSt = c.to.station != null ? c.to.station.name : "";
            System.out.printf("%-3d %s %-10s %s %-10s %-10s %-8d%n",
                i+1, depTime, depSt, arrTime, arrSt, c.duration, c.transfers);
        }
        if (sorted.size() > limit) {
            System.out.printf("... and %d more%n", sorted.size() - limit);
        }
    }

    private static void exportJSON(List<Connection> connections, String filename) throws IOException {
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        String json = gson.toJson(connections);
        try (FileWriter fw = new FileWriter(filename)) {
            fw.write(json);
        }
        System.out.println("Exported to " + filename);
    }

    private static void exportCSV(List<Connection> connections, String filename) throws IOException {
        try (PrintWriter pw = new PrintWriter(filename)) {
            pw.println("Departure,Departure Station,Arrival,Arrival Station,Duration,Changes");
            for (Connection c : connections) {
                pw.printf("%s,\"%s\",%s,\"%s\",%s,%d%n",
                    c.from.departure, c.from.station.name,
                    c.to.arrival, c.to.station.name,
                    c.duration, c.transfers);
            }
        }
        System.out.println("Exported to " + filename);
    }

    private static void interactive() throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        System.out.println("=== Train Schedule ===");
        System.out.print("From station: ");
        String from = reader.readLine().trim();
        if (from.isEmpty()) { System.out.println("Station required."); return; }
        System.out.print("To station: ");
        String to = reader.readLine().trim();
        if (to.isEmpty()) { System.out.println("Station required."); return; }
        System.out.print("Date (YYYY-MM-DD, leave blank for today): ");
        String date = reader.readLine().trim();
        if (date.isEmpty()) date = LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE);
        System.out.print("Time (HH:MM, leave blank for now): ");
        String time = reader.readLine().trim();
        if (time.isEmpty()) time = LocalDateTime.now().format(DateTimeFormatter.ofPattern("HH:mm"));
        List<Connection> conns = fetchConnections(from, to, date, time);
        display(conns, 10, "departure");
        if (!conns.isEmpty()) {
            System.out.print("\nExport results? (y/n): ");
            String exp = reader.readLine().trim().toLowerCase();
            if (exp.equals("y")) {
                System.out.print("Format (json/csv): ");
                String fmt = reader.readLine().trim().toLowerCase();
                System.out.print("Filename (default: schedule." + fmt + "): ");
                String fname = reader.readLine().trim();
                if (fname.isEmpty()) fname = "schedule." + fmt;
                if (fmt.equals("json")) exportJSON(conns, fname);
                else if (fmt.equals("csv")) exportCSV(conns, fname);
                else System.out.println("Unknown format.");
            }
        }
    }

    public static void main(String[] args) throws Exception {
        if (args.length > 0) {
            Map<String, String> params = new HashMap<>();
            for (int i = 0; i < args.length; i++) {
                if (args[i].startsWith("--")) {
                    String key = args[i].substring(2);
                    if (i + 1 < args.length && !args[i+1].startsWith("--")) {
                        params.put(key, args[++i]);
                    } else {
                        params.put(key, "true");
                    }
                }
            }
            if (!params.containsKey("from") || !params.containsKey("to")) {
                System.err.println("--from and --to are required.");
                return;
            }
            String from = params.get("from");
            String to = params.get("to");
            String date = params.getOrDefault("date", null);
            String time = params.getOrDefault("time", null);
            int limit = Integer.parseInt(params.getOrDefault("limit", "10"));
            String sortBy = params.getOrDefault("sort", "departure");
            String exportFmt = params.getOrDefault("export", null);
            String output = params.getOrDefault("output", null);
            List<Connection> conns = fetchConnections(from, to, date, time);
            display(conns, limit, sortBy);
            if (exportFmt != null && !conns.isEmpty()) {
                String fname = output != null ? output : "schedule." + exportFmt;
                if (exportFmt.equals("json")) exportJSON(conns, fname);
                else if (exportFmt.equals("csv")) exportCSV(conns, fname);
            }
        } else {
            interactive();
        }
    }
}
